drivers/ethernet/intel/eth_intel_igc.c

```diff
@@ -0,0 +1,1363 @@
+		      timeout, k_msleep(1))) {
+		LOG_ERR("Pending for GIO Master Request");
+		return -EIO;
+	}
+
+	return 0;
+}
+
+static void eth_intel_igc_init_speed(struct eth_intel_igc_mac_data *data)
+{
+	mm_reg_t base = data->base;
+
+	igc_modify(base, INTEL_IGC_CTRL, INTEL_IGC_CTRL_FRCSPD | INTEL_IGC_CTRL_FRCDPX, false);
+	igc_modify(base, INTEL_IGC_CTRL, INTEL_IGC_CTRL_SLU, true);
+}
+
+static void eth_intel_igc_get_dev_ownership(struct eth_intel_igc_mac_data *data)
+{
+	igc_modify(data->base, INTEL_IGC_CTRL_EXT, INTEL_IGC_CTRL_EXT_DRV_LOAD, true);
+}
+
+static int eth_intel_igc_init_mac_hw(const struct device *dev)
+{
+	struct eth_intel_igc_mac_data *data = dev->data;
+	int ret = 0;
+
+	ret = eth_intel_igc_disable_pcie_master(data->base);
+	if (ret < 0) {
+		LOG_ERR("GIO master request failed");
+		return ret;
+	}
+
+	igc_write((data->base + INTEL_IGC_IMC), UINT32_MAX);
+	igc_write((data->base + INTEL_IGC_RCTL), 0);
+	igc_write((data->base + INTEL_IGC_TCTL), INTEL_IGC_TCTL_PSP);
+	igc_reg_refresh(data->base);
+	k_msleep(20);
+
+	/* MAC Reset */
+	igc_modify(data->base, INTEL_IGC_CTRL, INTEL_IGC_CTRL_DEV_RST, true);
+
+	/* MAC receive address Init */
+	eth_intel_igc_rx_addrs_init(dev);
+
+	eth_intel_igc_get_dev_ownership(data);
+
+	return ret;
+}
+
+static int eth_intel_igc_init(const struct device *dev)
+{
+	const struct eth_intel_igc_mac_cfg *cfg = dev->config;
+	struct eth_intel_igc_mac_data *data = dev->data;
+	int ret = 0;
+
+	data->mac = dev;
+	data->base = DEVICE_MMIO_GET(cfg->platform);
+	if (!data->base) {
+		LOG_ERR("Failed to get MMIO base address");
+		return -ENODEV;
+	}
+
+	ret = eth_intel_igc_init_mac_hw(dev);
+	if (ret < 0) {
+		return ret;
+	}
+
+	ret = eth_intel_igc_pcie_msix_setup(dev);
+	if (ret < 0) {
+		return ret;
+	}
+
+	ret = eth_intel_igc_tx_dma_init(dev);
+	if (ret < 0) {
+		return ret;
+	}
+
+	return eth_intel_igc_rx_dma_init(dev);
+}
+
+static const struct ethernet_api eth_api = {
+	.iface_api.init = eth_intel_igc_iface_init,
+	.get_capabilities = eth_intel_igc_get_caps,
+	.set_config = eth_intel_igc_set_config,
+	.send = eth_intel_igc_tx_data,
+	.get_phy = eth_intel_igc_get_phy,
+#if defined(CONFIG_NET_STATISTICS_ETHERNET)
+	.get_stats = eth_intel_igc_get_stats,
+#endif
+};
+
+#define NUM_QUEUES(n)  DT_INST_PROP(n, num_queues)
+#define NUM_TX_DESC(n) DT_INST_PROP(n, num_tx_desc)
+#define NUM_RX_DESC(n) DT_INST_PROP(n, num_rx_desc)
+#define NUM_MSIX(n)    NUM_QUEUES(n)
+
+#define INTEL_IGC_MAC_CONFIG_IRQ(n)                                                                \
+	static void eth##n##_cfg_irq(const struct device *dev)                                     \
+	{                                                                                          \
+		eth_intel_igc_map_intr_to_vector(dev);                                             \
+		eth_intel_igc_init_queue_work_##n(dev);                                            \
+		eth_intel_igc_init_speed(dev->data);                                               \
+	}
```

Comment:
eth_intel_igc_map_intr_to_vector(dev);
and
eth_intel_igc_init_speed(dev->data);
should be called outside this instance specific function